Adding new role management

makirill opened the pull request:
New role management is checking runtime config to create additional
accounts based on 'additonal_groups` parameter.

This allows to create two (admin and generic) accounts for users with
privileged role like cluster administrators.

makirill committed:
Adding new role management

New role management is checking runtime config to create additional
accounts based on 'additonal_groups` parameter.

This allows to create two (admin and generic) accounts for users with
privileged role like cluster administrators.

makirill committed:
Fixing user-reset functionality

The user-reset functionality is partially broken now since list for user
to reset is not including accounts that was created for the additional
groups configured in runtime.

This fix includes all accounts associated with user specified under
queue/user-management-*.json files.

makirill committed:
Merge branch 'master' into feature/user-roles

makirill committed:
Minor fixes after testing

oybed commented on inventory-generation/identity-management/main.yml:
```diff
@@ -13,6 +13,10 @@
       include_vars:
         file: "{{ directory }}/engagement.json"
 
+    - name: Read Runtime Data
```
Since "older" repos don't necessarily have the `runtime.json` file, we may want to add a check above and error out if it doesn't exist - see example above

makirill replied:
I've added check for the `runtime.json` file and failing the playbook if it's not exist (be7fc14a1cf9ee1db5732eba8f1e21e79d7c2760). Probably failing is not a good option but I thing most of the engagements already since change for `runtime.json` file was deployed to production couple weeks ago.

The problem is what should be used for `user_options` and `user_role` variables as default values. Right now `residency` and `do500` having different values for `user_role` and these values important for the admin role.

makirill replied:
Added checks for user_options.user_roles.options (7fc8de944913224f7da12e8354ab8c7fd4ac2d82)

oybed commented on inventory-generation/identity-management/main.yml:
```diff
@@ -30,27 +34,32 @@
+    - include_tasks: process_groups.yml
+      loop_control:
+        loop_var: group_item
+      loop: "{{ user_options.user_roles.options }}"
```
need some default values defined in case the `user_options` and/or `user_roles` isn't defined

makirill replied:
These values are different for `residency` and `do500` engagements and I am not sure if we need to hardcore it. Probably it will be better to fix lack of the `runtime.json` file in the engagement repository.

oybed commented on inventory-generation/identity-management/main.yml:
```diff
@@ -30,27 +34,32 @@
       set_fact:
-        users: "{{ (users | default([])) + [ { 'first_name': 'LDAP', 'last_name': 'SA', 'email': 'ldap@localhost.com', 'user_name': ocp_ldap_sa_username, 'generate_password': False, 'notify_user': False, 'password': ocp_ldap_sa_password  } ] }}"
+        users: "{{ user_list.keys() | map('extract', user_list) | list | flatten | unique }}"
+      when: users is defined
```
`when: users is defined` will never be true from what I can tell - did you mean to check for `user_list` here?

makirill replied:
Yes, it should be `user_list` and I am wander how it worked for `users` variable. Fixed in be7fc14a1cf9ee1db5732eba8f1e21e79d7c2760

oybed commented on inventory-generation/identity-management/main.yml:
```diff
@@ -30,27 +34,32 @@
       set_fact:
-        usrgrp: "{{ (usrgrp | default([])) + [ {'name': item, 'members': (engagement_users | selectattr('role','equalto',item) | map(attribute='email') | map('regex_replace','@.*','') | list) } ] }}"
-      loop: "{{ unique_groups }}"
+        unique_groups: "{{ user_options.user_roles.options | json_query('[].value') | unique }}"
```
same as above - need default values in case undefined

oybed commented on inventory-generation/identity-management/queue/user-management.yml:
```diff
@@ -4,15 +4,7 @@
   set_fact:
-    users_remove: "{{ (users_remove | default([])) + [ user_info ] }}"
+    users_remove: "{{ users_remove|d([]) + [ item | combine({'state': 'absent'})] }}"
+  loop: "{{ users | selectattr('email', 'equalto', email ) | list }}"
```
No sure I follow the logic here. Two thnings:
1) the `users` list isn't the correct one to use here
2) we still need the logic above to correctly generate the `user_name` based on e-mail.

... and 3) - which may have been an issue in the past as well - we should be able to process a list of users. IIRC the queue job file may include multiple users (I may remember incorrectly on this, but something to check), but if that's not the case we should be good as-is - just need to correct 1) and 2) above

makirill replied:
How it works now. The `lodestar-frontend` generates files in the `queue` folder with user dictionary which is exactly the same as we can find in `engagement.json`. What is necessary is to get all users from all files in `queue` folder, add `state: absent` field and put it under `lodestar_identities_remove:` at `ocp-init/id-mgmt.yaml` file. 

The `lodestar-frontend` don't have information about situations when user could have two or more accounts like generic and admin one, so `queue` folder will only contain information about one single account but we need to reset passwords to all the accounts. The common thing about all the accounts is email, so I am creating list of the users for `lodestar_identities_remove:` to reset all accounts associated with particular email. The `users` list is having all the current accounts from `engagement.json` and filtering by email address generates proper list for `users_remove`

makirill replied:
Remove Queue generation recreated (90874c5ee03226bc52eb48fc28663b75c1cb3659)

oybed reviewed (changes requested):
See inline comments. I'd be happy to jump on a call to discuss if needed

makirill committed:
Adding check for runtime.json file

makirill committed:
Fixing remove queue generation

makirill committed:
Adding verification to user_options.user_roles.options

oybed committed:
Code fine-tuning and clean-up

oybed committed:
Code fine-tuning and clean-up

oybed committed:
Code fine-tuning and clean-up

oybed committed:
Code fine-tuning and clean-up

makirill committed:
Merge pull request #1 from oybed/user-roles-cleanup

Code fine-tuning and clean-up

makirill committed:
Minor fixes

- process_list_of_users.yml called twice, so user_list variable contain
  data not only form list of users to be added but list of users to
remove (queue). This leads to incorrect group information.

- state absent was missed from users listed for removal

oybed committed:
Removing debug code

Removing debug code

oybed reviewed (approved):
LGTM
